pkg/placement/workload-projector.go

```diff
@@ -141,89 +205,175 @@ func (wp *workloadProjector) configSyncLoop(ctx context.Context, worker int) {
+				},
+				Spec: wp.syncerConfigSpecFromRelations(goodConfigSpecRelations)}
+			client := wp.edgeClusterClientset.EdgeV1alpha1().Cluster(scRef.Cluster.Path()).SyncerConfigs()
+			syncfg2, err := client.Create(ctx, syncfg, metav1.CreateOptions{FieldManager: "placement-translator"})
+			if logger.V(4).Enabled() {
+				logger = logger.WithValues("specNamespaces", syncfg.Spec.NamespaceScope.Namespaces,
+					"specResources", syncfg.Spec.NamespaceScope.Resources,
+					"specClusterObjects", syncfg.Spec.ClusterScope)
+			}
+			if err == nil {
+				logger.V(2).Info("Created SyncerConfig", "resourceVersion", syncfg2.ResourceVersion)
+				return false
+			}
+			logger.Error(err, "Failed to create SyncerConfig")
+			return true
+		}
+		logger.Error(err, "Unexpected failure reading local cache")
+	}
+	goodConfigSpecRelations := wp.syncerConfigRelations(sp)
+	if wp.syncerConfigIsGood(sp, scRef, syncfg, goodConfigSpecRelations) {
+		logger.V(4).Info("SyncerConfig is already good", "resourceVersion", syncfg.ResourceVersion)
+		return false
+	}
+	syncfg.Spec = wp.syncerConfigSpecFromRelations(goodConfigSpecRelations)
+	client := wp.edgeClusterClientset.EdgeV1alpha1().Cluster(scRef.Cluster.Path()).SyncerConfigs()
+	syncfg2, err := client.Update(ctx, syncfg, metav1.UpdateOptions{FieldManager: "placement-translator"})
+	if logger.V(4).Enabled() {
+		logger = logger.WithValues("specNamespaces", syncfg.Spec.NamespaceScope.Namespaces,
+			"specResources", syncfg.Spec.NamespaceScope.Resources,
+			"specClusterObjects", syncfg.Spec.ClusterScope)
+	}
+	if err != nil {
+		logger.Error(err, "SyncerConfig update failed", "resourceVersion", syncfg.ResourceVersion)
+		return true
+	}
+	logger.V(2).Info("Updated SyncerConfig", "resourceVersionOld", syncfg.ResourceVersion, "resourceVersionNew", syncfg2.ResourceVersion)
 	return false
 }
 
 func (wp *workloadProjector) Transact(xn func(WorkloadProjectionSections)) {
 	logger := klog.FromContext(wp.ctx)
 	wp.Lock()
 	defer wp.Unlock()
+	logger.V(3).Info("Begin transaction")
 	var s1 SetChangeReceiver[Triple[edgeapi.SinglePlacement, NamespaceName, logicalcluster.Name]] = wp.nsDistributions
 	var s2 SetChangeReceiver[Triple[edgeapi.SinglePlacement, metav1.GroupResource, logicalcluster.Name]] = wp.nsrDistributions
 	var s3 SetChangeReceiver[Triple[edgeapi.SinglePlacement, GroupResourceInstance, logicalcluster.Name]] = wp.nnsDistributions
-	changedDestinations := NewMapSet[edgeapi.SinglePlacement]()
-	wp.changedDestinations = &changedDestinations
-	s1 = SetChangeReceiverFork(false, s1, recordFirst[edgeapi.SinglePlacement, NamespaceName, logicalcluster.Name](changedDestinations))
-	s2 = SetChangeReceiverFork(false, s2, recordFirst[edgeapi.SinglePlacement, metav1.GroupResource, logicalcluster.Name](changedDestinations))
-	s3 = SetChangeReceiverFork(false, s3, recordFirst[edgeapi.SinglePlacement, GroupResourceInstance, logicalcluster.Name](changedDestinations))
+	changedDestinations := func() *MutableSet[edgeapi.SinglePlacement] {
```

Comment:
Looks like there're quite a few usages of anonymous functions in recent PRs. Here is the 1st example, and there are a [2nd one](https://github.com/kcp-dev/edge-mc/blob/05b7f6c0f0a9e3398fc4b8a26dc14a28e8338597/pkg/placement/apiwatch-map-provider.go#L231-L244) a [3rd one](https://github.com/kcp-dev/edge-mc/blob/05b7f6c0f0a9e3398fc4b8a26dc14a28e8338597/pkg/placement/apiwatch-map-provider.go#L271-L276). I have two questions on them:
1. For the 2nd one and the 3rd one, I guess the anonymous functions are used to specify the location for the `defer` statements, am I right?
2. For the 1st anonymous function here, I'm trying to understand, what are the benefits or it is just some personal style?

Reply:
The 1st one here is the scars from a battle over mysterious loss of information in the MutableSet here.  I tried making an incontrovertibly shared pointer, thinking maybe different parts of the code were manipulating different copies of the set.  The battle was finally won by wrapping the set with a Mutex, and I have not gotten around to clearing out the junk.  Will attend to in the future.

The other two examples are, as you observed, cases where I want the mutex to apply to just a small section of the code.

Reply:
I see. Thanks for explaining this!